### PR TITLE
socok8s: add ansible lint job

### PR DESCRIPTION
**Pull request by Itxaka**
Adds a job for ansible linting the socok8s playbooks

Lint file is provided on https://github.com/SUSE-Cloud/socok8s/pull/331

### CODE REVIEW

File: jenkins/ci.suse.de/cloud-socok8s.yaml [hunk 1]
```diff
@@ -6,3 +6,4 @@
     repo-credentials: c2350527-476a-45df-b406-84f028614682
     jobs:
       - '{name}-integration'
+      - '{name}-ansible-lint'
```
Comment (toabctl) [inline jenkins/ci.suse.de/cloud-socok8s.yaml]: I would prefer to call it just `lint`. we maybe add more linting beside the ansible one.

File: jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+          repo: '{repo-name}'
+          repo-owner: '{repo-owner}'
+          credentials-id: '{repo-credentials}'
+          branch-discovery: no-pr
+          discover-pr-forks-strategy: merge-current
+          discover-pr-forks-trust: permission
+          discover-pr-origin: merge-current
+          submodule:
+            recursive: true
+          notification-context: continuous-integration/jenkins
```
Comment (toabctl) [inline jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml]: not sure, but shouldn't that be something like `continuous-integration/jenkins/lint` ?

File: jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    number-to-keep: 30
+    days-to-keep: 30
+    script-path: Jenkinsfile.lint
+    scm:
+      - github:
+          repo: '{repo-name}'
+          repo-owner: '{repo-owner}'
+          credentials-id: '{repo-credentials}'
+          branch-discovery: no-pr
+          discover-pr-forks-strategy: merge-current
```
Comment (toabctl) [inline jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml]: I think `current` is enough .AFAIU, `merge-current` results in rechecks whenever the target branch (in our case master) changes. But for linting, that is not needed imo.

File: jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+          discover-pr-forks-trust: permission
+          discover-pr-origin: merge-current
```
Comment (toabctl) [inline jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml]: same here. use `current`

File: jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+- job-template:
+    name: '{name}-ansible-lint'
```
Comment (toabctl) [inline jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml]: same here.

File: jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+- job-template:
```
Comment (jhesketh) [inline jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml]: (nit) I would have just put this in the same file as the other socok8s tests (although that should probably be renamed from integrations)

File: jenkins/ci.suse.de/cloud-socok8s.yaml (hunk 1 above)
Comment (Itxaka) [reply jenkins/ci.suse.de/cloud-socok8s.yaml]: I wasn wondering if we would create more than one linter and if it was on the same job or on a different one.... this seems to hint to the same job, Im ok with that :D

File: jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml (hunk 2 above)
Comment (Itxaka) [reply jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml]: ah indeed!

File: jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml (hunk 6 above)
Comment (Itxaka) [reply jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml]: can we? I dont know how :D

File: jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml (hunk 3 above)
Comment (Itxaka) [reply jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml]: done

File: jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml (hunk 6 above)
Comment (jhesketh) [reply jenkins/ci.suse.de/templates/cloud-socok8s-ansible-lint.yaml]: Just another entry in the file. I was about to point you to another file to look at, but it looks like they are all separate anyway, so this follows convention :-).

**Comment by Itxaka** [comment]
> The only thing I would like to know before merging, is the ability to stop the integrations job if the linters fail. If this needs to be in a single file to make it happen, then we should decide first if killing other jobs is something we want to achieve or not. From that, we would know if this should be merged as is, or need adaptation.

That is a really good point, Im not sure if there is a way of linking pipelines or we should just include an extra step on the existing pipeline.

Including the linting part in the existing pipeline would be the easiest way but....we would get a generic error on the PR. On the other side, if we fail the linting the job is immediately failed, which frees resources.

Should we just do that? Make it part of the normal pipeline? @evrardjp @toabctl @jhesketh

**Comment by jhesketh** [comment]
I agree, we should stop the integration tests to save on resources if linting fails.

The easiest way would be to have it as a separate step in the pipeline. This obviously makes the integration job longer and doesn't give us any parallel advantages. This is likely okay given the current runtime of the linting, but it may become an issue if that grows too large.

You can have a separate linting job cancel any running integration jobs for a PR. It's not too difficult to do (just a call to a jenkins URL), but the linting job would need to determine any integration job build id's somehow. That part is trickier (but not undoable) unless the linting job launches the integration job but I would like to avoid that as it somehow seems more complex.

An alternative though is to do it as a parallel stage in the one integration pipeline with failfast set (https://jenkins.io/doc/book/pipeline/syntax/#parallel). This is likely the easiest option, and even though it's all in one job, you can still see an overview of the status in the blueocean UI.

**Comment by Itxaka** [comment]
> I agree, we should stop the integration tests to save on resources if linting fails.

>The easiest way would be to have it as a separate step in the pipeline. This obviously makes the integration job longer and doesn't give us any parallel advantages. This is likely okay given the current runtime of the linting, but it may become an issue if that grows too large.

I think this is the easiest and the fastest way of doing it and as you say, its currently fast enough that we dont need to run it on parallel so I'll do this.
If it becomes too large or too slow then I guess we can tackle that.

>You can have a separate linting job cancel any running integration jobs for a PR. It's not too difficult to do (just a call to a jenkins URL), but the linting job would need to determine any integration job build id's somehow. That part is trickier (but not undoable) unless the linting job launches the integration job but I would like to avoid that as it somehow seems more complex.

>An alternative though is to do it as a parallel stage in the one integration pipeline with failfast set (jenkins.io/doc/book/pipeline/syntax/#parallel). This is likely the easiest option, and even though it's all in one job, you can still see an overview of the status in the blueocean UI.

But this would still create resources that may be cancelled afterwards no? I guess that if we want to avoid creating uneeded resources and jobs taking too long (cleanup also takes time!) we should avoid this.


Closing this and adding it on the jenkinsfile directly! thanks for all the comments and hints!

**Pull request closed without merging**